python/cudf/cudf/core/dtype_matching.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+from __future__ import annotations
+
+import functools
+from typing import Tuple, cast
+
+import numpy as np
+
+import cudf.core.column as column
+from cudf.core.dtypes import is_categorical_dtype
+
+
+@functools.cache
+def common_numeric_type(ltype: np.dtype, rtype: np.dtype) -> np.dtype | None:
```

Comment:
TODO: Write a test for this that asserts that the full promotion table generated by this scheme is a partial semi-lattice. That way we assure ourselves that promotion is associative.